### ISSUE TITLE
Update fails with "Waiting for app ReplicaSet be marked available"

### ISSUE THREAD

**Issue by ncsibra** [pulumi/pulumi-kubernetes#497]
Hi,

I have the same issue as here: #485 

```
[1/2] Waiting for app ReplicaSet be marked available
Plan apply failed: 2 errors occurred:

* Timeout occurred for 'sbx-chartmuseum-chartmuseum'
* Minimum number of Pods to consider the application live was not attained
```

The requested outputs from the other issue:
```
➜  k8s-chartmuseum (80-chartmuseum) ✗ kubectl get rs -n deployment
NAME                                     DESIRED   CURRENT   READY   AGE
sbx-chartmuseum-chartmuseum-78df45b986   1         1         1       2h
sbx-chartmuseum-chartmuseum-867bf555b8   0         0         0       5h
➜  k8s-chartmuseum (80-chartmuseum) ✗ kubectl get deploy -n deployment
NAME                          DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
sbx-chartmuseum-chartmuseum   1         1         1            1           5h
```

The trimmed resource properties from pulumi dashboard:
```
{
    "apiVersion": "extensions/v1beta1",
    "kind": "Deployment",
    "metadata": {
        "annotations": {
            "deployment.kubernetes.io/revision": "2"
        },
        "creationTimestamp": "2019-03-22T07:51:54Z",
        "generation": 2,
        "labels": {
            "app": "chartmuseum",
            "app.kubernetes.io/managed-by": "pulumi",
            "chart": "chartmuseum-2.1.0",
            "heritage": "Tiller",
            "release": "sbx-chartmuseum"
        },
        "name": "sbx-chartmuseum-chartmuseum",
        "namespace": "deployment",
        "resourceVersion": "2327596",
        "selfLink": "/apis/extensions/v1beta1/namespaces/deployment/deployments/sbx-chartmuseum-chartmuseum",
        "uid": "5c244ea1-4c77-11e9-bc03-0a51bd93144a"
    },
    "spec": {
        "progressDeadlineSeconds": 2147483647,
        "replicas": 1,
        "revisionHistoryLimit": 10,
        "selector": {
            "matchLabels": {
                "app": "chartmuseum",
                "release": "sbx-chartmuseum"
            }
        },
        "strategy": {
            "rollingUpdate": {
                "maxSurge": 1,
                "maxUnavailable": 1
            },
            "type": "RollingUpdate"
        },
        "template": {
            "metadata": {
                "labels": {
                    "app": "chartmuseum",
                    "release": "sbx-chartmuseum"
                },
                "name": "sbx-chartmuseum-chartmuseum"
            },
            "spec": {
                "containers": [
                    {
                        "args": [
                        ],
                        "env": [
                        ],
                        "image": "chartmuseum/chartmuseum:v0.8.2",
                        "imagePullPolicy": "IfNotPresent",
                        "livenessProbe": {
                            "failureThreshold": 3,
                            "httpGet": {
                                "path": "/health",
                                "port": "http",
                                "scheme": "HTTP"
                            },
                            "initialDelaySeconds": 5,
                            "periodSeconds": 10,
                            "successThreshold": 1,
                            "timeoutSeconds": 1
                        },
                        "name": "chartmuseum",
                        "ports": [
                            {
                                "containerPort": 8080,
                                "name": "http",
                                "protocol": "TCP"
                            }
                        ],
                        "readinessProbe": {
                            "failureThreshold": 3,
                            "httpGet": {
                                "path": "/health",
                                "port": "http",
                                "scheme": "HTTP"
                            },
                            "initialDelaySeconds": 5,
                            "periodSeconds": 10,
                            "successThreshold": 1,
                            "timeoutSeconds": 1
                        },
                        "resources": {},
                        "terminationMessagePath": "/dev/termination-log",
                        "terminationMessagePolicy": "File"
                    }
                ],
                "dnsPolicy": "ClusterFirst",
                "nodeSelector": {
                    "dliver.com/stack": "deployment"
                },
                "restartPolicy": "Always",
                "schedulerName": "default-scheduler",
                "securityContext": {
                    "fsGroup": 1000
                },
                "terminationGracePeriodSeconds": 30,
                "volumes": [
                    {
                        "emptyDir": {},
                        "name": "storage-volume"
                    }
                ]
            }
        }
    },
    "status": {
        "availableReplicas": 1,
        "conditions": [
            {
                "lastTransitionTime": "2019-03-22T07:51:54Z",
                "lastUpdateTime": "2019-03-22T07:51:54Z",
                "message": "Deployment has minimum availability.",
                "reason": "MinimumReplicasAvailable",
                "status": "True",
                "type": "Available"
            }
        ],
        "observedGeneration": 2,
        "readyReplicas": 1,
        "replicas": 1,
        "updatedReplicas": 1
    }
}
```

I assumed the problem is, that pulumi waiting for the old ReplicaSet too, which has 0 desired capacity, but after manually deleted the old ReplicaSet, pulumi still failed with the same error.

**Comment by lblackstone** [comment]
Hi @ncsibra, I'll need some more information to diagnose what's going on here. Can you run `kubectl describe deploy sbx-chartmuseum-chartmuseum -n deployment` and see if the events give some hints?

I don't think the problem is related to the old ReplicaSet.

**Comment by ncsibra** [comment]
There is no event for that deployment.

```
Name:                   sbx-chartmuseum-chartmuseum
Namespace:              deployment
CreationTimestamp:      Fri, 22 Mar 2019 08:51:54 +0100
Labels:                 app=chartmuseum
                        app.kubernetes.io/managed-by=pulumi
                        chart=chartmuseum-2.1.0
                        heritage=Tiller
                        release=sbx-chartmuseum
Annotations:            deployment.kubernetes.io/revision: 2
Selector:               app=chartmuseum,release=sbx-chartmuseum
Replicas:               1 desired | 1 updated | 1 total | 1 available | 0 unavailable
StrategyType:           RollingUpdate
MinReadySeconds:        0
RollingUpdateStrategy:  1 max unavailable, 1 max surge
Pod Template:
  Labels:  app=chartmuseum
           release=sbx-chartmuseum
  Containers:
   chartmuseum:
    Image:      chartmuseum/chartmuseum:v0.8.2
    Port:       8080/TCP
    Host Port:  0/TCP
    Args:
    Liveness:   http-get http://:http/health delay=5s timeout=1s period=10s #success=1 #failure=3
    Readiness:  http-get http://:http/health delay=5s timeout=1s period=10s #success=1 #failure=3
    Environment:
    Mounts:                        <none>
  Volumes:
   storage-volume:
    Type:    EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:  
Conditions:
  Type           Status  Reason
  ----           ------  ------
  Available      True    MinimumReplicasAvailable
OldReplicaSets:  <none>
NewReplicaSet:   sbx-chartmuseum-chartmuseum-78df45b986 (1/1 replicas created)
Events:          <none>
```

**Comment by ncsibra** [comment]
Btw, it's unable to refresh the state of the deployment, even when it was successful and the application working fine.

```
error: Preview failed: 2 errors occurred:
    
    * Resource 'sbx-chartmuseum-chartmuseum' was created but failed to initialize
    * Minimum number of Pods to consider the application live was not attained
```

**Comment by hausdorff** [comment]
@ncsibra can you show us the `kubectl get -o yaml` of that deployment resource? I'm curious to see what the last status was.

**Comment by hausdorff** [comment]
Oh, sorry, you already did... one second. :)

**Comment by hausdorff** [comment]
@ncsibra what version of Kubernetes are you running? Typically when a rollout succeeds, we expect the `.status` field to be updated with a second condition indicating that the replicaset rollout was successful. It looks like this does not happen here -- all we see is that it's available, and so we conclude that rollout was _not_ successful. So I'm wondering if some very old versions of the `Deployment` controller do not provide this condition in the `.status` field.

**Comment by ncsibra** [comment]
We're using AWS EKS, 1.11, platform version eks.2.
```
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.0", GitCommit:"641856db18352033a0d96dbc99153fa3b27298e5", GitTreeState:"clean", BuildDate:"2019-03-25T15:53:57Z", GoVersion:"go1.12.1", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"11+", GitVersion:"v1.11.8-eks-7c34c0", GitCommit:"7c34c0d2f2d0f11f397d55a46945193a0e22d8f3", GitTreeState:"clean", BuildDate:"2019-03-01T22:49:39Z", GoVersion:"go1.10.8", Compiler:"gc", Platform:"linux/amd64"}
```
What kind of condition do you expect exactly? Something like this?
```
status:
  availableReplicas: 1
  conditions:
  - lastTransitionTime: "2019-03-21T10:33:22Z"
    lastUpdateTime: "2019-03-21T10:33:25Z"
    message: ReplicaSet "prometheus-operator-operator-6969d5cfdf" has successfully
      progressed.
    reason: NewReplicaSetAvailable
    status: "True"
    type: Progressing
  - lastTransitionTime: "2019-03-22T08:02:26Z"
    lastUpdateTime: "2019-03-22T08:02:26Z"
    message: Deployment has minimum availability.
    reason: MinimumReplicasAvailable
    status: "True"
    type: Available
  observedGeneration: 1
  readyReplicas: 1
  replicas: 1
  updatedReplicas: 1
```
This chart was deployed with pulumi too, to the same EKS.
The funny thing is, that we deployed the chartmuseum to our dev environment too, it has the same status condition as I pasted(only that one) and pulumi was able to recognize it as ready. :)
I can try to delete and redeploy the whole stack and see if we can reproduce this again. I only kept in this state, you may need more information.

**Comment by hausdorff** [comment]
Yes -- for generations >= 1, when a replica set progresses, it `NewReplicaSetAvailable` should be `"True"`. For generation 1, my understanding (which may be wrong) is that this status is not present because the controller has not rolled out a new replica set yet -- only created. In your case, I see that it's generation 2 so I would expect this condition to be correct.

Would it be possible for you to send me your exact EKS and Chart code? If not on here, happy to help from alex@pulumi.com. Thanks!

**Comment by ncsibra** [comment]
Now the same thing happens with tiller, which is created from this yaml:

```yaml
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: tiller
  namespace: kube-system

---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRoleBinding
metadata:
  name: tiller-role-binding
roleRef:
  kind: ClusterRole
  name: cluster-admin
  apiGroup: rbac.authorization.k8s.io
subjects:
- kind: ServiceAccount
  name: tiller
  namespace: kube-system

---
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  creationTimestamp: null
  labels:
    app: helm
    name: tiller
  name: tiller-deploy
  namespace: kube-system
spec:
  replicas: 1
  strategy: {}
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: helm
        name: tiller
    spec:
      automountServiceAccountToken: true
      serviceAccountName: tiller
      containers:
      - env:
        - name: TILLER_NAMESPACE
          value: kube-system
        - name: TILLER_HISTORY_MAX
          value: "0"
        image: gcr.io/kubernetes-helm/tiller:v2.13.1
        imagePullPolicy: IfNotPresent
        livenessProbe:
          httpGet:
            path: /liveness
            port: 44135
          initialDelaySeconds: 1
          timeoutSeconds: 1
        name: tiller
        ports:
        - containerPort: 44134
          name: tiller
        - containerPort: 44135
          name: http
        readinessProbe:
          httpGet:
            path: /readiness
            port: 44135
          initialDelaySeconds: 1
          timeoutSeconds: 1
        resources: {}
status: {}

---
apiVersion: v1
kind: Service
metadata:
  creationTimestamp: null
  labels:
    app: helm
    name: tiller
  name: tiller-deploy
  namespace: kube-system
spec:
  ports:
  - name: tiller
    port: 44134
    targetPort: tiller
  selector:
    app: helm
    name: tiller
  type: ClusterIP
status:
  loadBalancer: {}

...
```
with `k8s.yaml.ConfigFile`.
It starts to be a real issue becauese this way our stacks stays in a failed state, when on the kubernetes side everything works fine.

**Comment by hausdorff** [comment]
This should be fixed as of `@pulumi/kubernetes` v0.23.1. `extensions/v1beta1` does not publish all the status messages that later API versions do, and so we "weakened" the await conditions, which should result in this working. @lblackstone is there a reason this is still open?

**Comment by lblackstone** [comment]
Yeah, this should have been fixed by #523 and included starting in the `v0.22.1` release.
